src/query/v2/physical/physical.hpp

```diff
@@ -0,0 +1,281 @@
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include <iostream>
+#include <memory>
+#include <vector>
+
+#include "query/plan/operator.hpp"
+#include "query/v2/physical/multiframe.hpp"
+#include "utils/thread_pool.hpp"
+#include "utils/visitor.hpp"
+
+namespace memgraph::query::v2::physical {
+
+using memgraph::query::plan::HierarchicalLogicalOperatorVisitor;
+using memgraph::query::plan::Once;
+using memgraph::query::plan::Produce;
+using memgraph::query::plan::ScanAll;
+using memgraph::utils::ThreadPool;
+
+/// Moving/copying Frames between operators is questionable because operators
+/// mostly operate on a single Frame value.
+///
+/// TODO(gitbuda): Consider adding PipelineSegment abstraction.
+///
+struct PipelineSegment {};
+
+/// Access to the thread and data pools should be given via the
+/// ExecutionContext.
+///
+struct ExecutionContext {
+  ThreadPool *thread_pool;
+};
+
+/// THREAD POOLS
+/// There might be the case we need many different thread pools (some of them
+/// might be the same):
+///   * Query    Pool -> Responsible for top-level query work
+///   * Operator Pool -> Responsible for the operator work (e.g. data moves)
+///   * IO       Pool -> Responsible for the interactions with the storage
+///   * CPU      Pool -> Responsible for CPU heavy operations like aggregations
+/// TODO(gitbuda): Design thread pools and allocators.
+///   * Thread pool has to support cooperative cancellation of specific tasks.
+
+/// TODO(gitbuda): Consider variant instead of virtual for the PhysicalOperator.
+/// NOTE: Frame[OutputSymbol] tells where a single operator should place values.
+///
+/// Each derived PhysicalOperator may be constructed with 0|1 function providing the data.
+/// Each PhysicalOperator has an arbitrary number of children AKA input operators.
+class PhysicalOperator {
+ public:
+  PhysicalOperator() = delete;
+  PhysicalOperator(const PhysicalOperator &) = default;
+  PhysicalOperator(PhysicalOperator &&) = default;
+  PhysicalOperator &operator=(const PhysicalOperator &) = default;
+  PhysicalOperator &operator=(PhysicalOperator &&) = default;
+  explicit PhysicalOperator(const std::string &name) : name_(name) {}
+  virtual ~PhysicalOperator() {}
+
+  void AddChild(std::shared_ptr<PhysicalOperator> child) { children_.push_back(child); }
+  std::string Name() const { return name_; }
+
+  /// Init/start execution.
+  virtual void Execute(ExecutionContext ctx) = 0;
+  /// Get data from children operators.
+  virtual const multiframe::Multiframe *NextRead() = 0;
+  virtual const multiframe::Multiframe *NextWrite() = 0;
```

Comment:
Somehow blocks (e.g. exponential backoff) if the underlying Multiframe is not accessible at the moment